pkg/proxy/oxyproxy.go

```diff
@@ -0,0 +1,607 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/common/addressing"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/nodeaddress"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/proxy/accesslog"
+
+	"github.com/braintree/manners"
+	log "github.com/sirupsen/logrus"
+	"github.com/vulcand/oxy/forward"
+	"github.com/vulcand/route"
+)
+
+// OxyRedirect implements the Redirect interface for a l7 proxy
+type OxyRedirect struct {
+	mutex lock.RWMutex
```

Comment:
what is this mutex protecting? @tgraf ?

Reply:
It is protecting the routes (== HTTP policy rules) that can be updated concurrently with policy enforcement.